src/App.js

```diff
@@ -1,10 +1,51 @@
-import React, { useState } from "react";
+import React, { useReducer, useState } from "react";
 import Navbar from "./components/Nav";
 import Container from "./components/Container";
+import Form from "./components/Form";
 import "../src/style/form.css";
 import "./style/index.css";
+import { useRecoilState, useSetRecoilState } from "recoil";
+import { todoAtom, searchAtom } from "./store/atom";
+import { type } from "@testing-library/user-event/dist/type";
 
-const initTodo = [
+// 추상황 -> 재사용성 up, 테스트하기 용이
+// 리스트, 검색상태, 조회상태, 수정상태
+// 리스트 = 할일을 보여주면 되고
+// 검색상태 = 리스트에 필터건 할일 보여주면 되고
+// 수정상태 = 선택한 할일 보여주면 되고
+
+// ! 이 방법이 베스트임. 함수 추상화 -> 함수형프로그래밍 -> 단점: 코드량 많아서
+// ui - 비지니스 로직 분리를 하자.
+//1. react-hooks(custom-hooks), 2.컴포넌트 단에서 나누는 방법
+const reducer = (todos, action) => {
+  switch (action.type) {
+    case 'ADD':
+      return [ ...todos, {
+        id: action.id + 1,
+        title: action.title, 
+        body: action.body 
+      }
+    ];
+    case 'EDIT':
+      const updatedTodos = [...todos];
+      const index = updatedTodos.findIndex((todo) => todo.id === action.editTodoId);
+      
+      updatedTodos[index] = {
+          id: action.editTodoId,
+          title: action.title,
+          body: action.body,
+      };
+      return updatedTodos;
+    // case 'SEARCH':
+    //   const searchResult = todos.filter((todo) => todo.title.toLowerCase()
+    //   .includes(action.searchTerm.toLowerCase()));
+    //   return searchResult;
+    default:
+      return todos;
+  }
```

Comment:
*lookup table* 이라는 엑셀에서 사용하는 데이터 자료형이 있습니다. key,value 구성된 미리 데이터를 정형화 시켜놓 테이블을 하나 더 빼놓은 거예요. 전날에 Vlookup 사용해보셨냐고 질문드렸었는데 Vlookup 함수를 사용해서 다른테이블을 만들잖아요? 그런겁니다.
**조건문 자체를 변수에 데이터로 저장한다고 생각됩니다..**
코드로 한번 해보면
```suggestion
  const copyTodos =  [...todos]
  const lookupTable = {
  'ADD': [ ...todos, {
        id: action.id + 1,
        title: action.title, 
        body: action.body 
        }
      ],
  'EDIT': copyTodos.find((todo) => todo.id === action.editTodoId),
  "DEFAULT": todos,
  }
  return lookupTable[['ADD', 'EDIT'].includes(action.type) ? action.type : 'DEFAULT']
```

이렇게 하게되면 만약 조건분기 기능 수정할떄 유지보수가 용이하고 코드도 알아보기 쉽습니다. 또한 느슨한 결합으로 **lookuptabel** 함수형 프로그래밍에 이점으로 테스트에 용이하고 유연한 함수로 만들 수 있습니다.